### DISCUSSION TITLE
Edited images not exporting correctly

### DISCUSSION THREAD

**Discussion by drodner** [RhetTbull/osxphotos#856] [Q&A]
Hi, 

I recently ran an export on a large photos library of roughly 155000 pictures. It's Photos 4.0 on MacOS Mojave.  I'm trying to export so I can migrate to digiKam, but I'm seeing a problem with the export of edited images.

In our Photos library, displaying only the Edited photos shows about 5900 pictures. When I ran the export, I only got 250 files with a filename including "_edited"

All of these 250, except for 2, were in a 2019/_ folder -- photos that were not in albums. However, we also have many edited photos that are in albums, and many edited photos from other years as well.

There are many photos that have duplicate file names, especially photos that are not in albums and thus export to an '_' folder. Lots of DSC_0001.JPG for example. Exported files correctly have DSC_0001 (1).JPG, DSC_0001 (2).JPG, etc.  But I think maybe edited files are being overwritten, or not all of them exported.  In all the cases I've looked at, the DSC_0001_edited.JPG is a very different picture than DSC_0001.JPG. But it does match with one of the DSC_0001 (n) images.  I think the edited version was exported without (or before) the original image got the (n) added to its filename, and so the few _edited images I have do not have correct filenames for their corresponding original images.

Wondering if there is a fix for this or if there is something I need to do differently.

Thanks for your help!

**Comment by oPromessa** [comment]
Hi @drodner   

I have a similar issue in which sometimes the edited pic is not found. And osxphotos exports the original. 

```log
2022-12-04 07:47:45.116430 -- Edited file for PC280055.JPG is missing, exporting original
```

I use the option on the `toml` file: `skip_original_if_edited = true`

My pics in Photos are all referenced, just FYI. 

Still investigating but using `osxphotos inspect` while reverting back to original and re-editing the pic (normally what I do is rotating them or red-eye removal) some times it shows the path of the **edited** pic and sometimes no. 

So I think either the database of Photos is messed up (in some cases i had to delete and re-upload the pic) or osxphotos is not able to find the path for the edited pic. 

Do check this topic #610 which may be helpful also. 

Thanks.

**Comment by RhetTbull** [comment]
What's the full command line you're using with osxphotos to do the export?  I'll see if I can replicate this. (Though I no longer have access to Mojave if this is a Mojave-specific bug)  Your hypothesis sounds plausible (though the code has been designed to avoid this so if it's happening, it's a bug).  It is true that osxphotos does not maintain the mapping between original and edited when there are duplicate filenames and this is something I cannot fix without a great deal of rework but it should not be overwriting edited images.  In the case where a duplicate original gets the `(1)`, `(2)` added and the second one (e.g. `DSC_0001 (1).jpg`) has been edited, the filename will be `DSC_0001_edited.jpeg`.  There are good reasons for this in the code and it's not an easy fix. 

One thing you could try would be to ensure each filename is unique.  There's an `{id}` template field that assigns a unique ID to each photo (an integer number). You could use the following `--filename` template to ensure unique names: `--filename "{original_name}_{id:06d}"`  The `:06d` formats the ID as a 6-digit integer with leading zeros (as you've got a library with 6 digits worth of photos).

In fact, it would be helpful if you could try a couple experiments to help debug this:

1. Select several Photos with duplicate names in Photos (e.g. search for `DSC_0001` in your example) including some with edits then run the export command adding `--selected` to export just the selected photos and see if you get the same problem.
2. Do the same thing again but using the `--filename` template from above.

**Comment by drodner** [comment]
My full command line was this:
osxphotos export "/Volumes/i\$/PhotosExport" --directory "{created.year}/{folder_album}" --jpeg-ext JPG --retry 3 --sidecar XMP --report export.csv --verbose 

I will try your two suggestions and report back.

**Reply by RhetTbull** [reply]
I tried to replicate this and was not able to.  I created a new library with several photos that had duplicate names. Edited some of these. Put some of them into albums. And my export was as expected: the edited photos were exported with the `(1)` etc. 

```
.
├── Test Album
│   ├── IMG_2510 (1).JPG
│   ├── IMG_2510 (1).JPG.xmp
│   ├── IMG_2510 (2).JPG
│   ├── IMG_2510 (2).JPG.xmp
│   ├── IMG_2510.JPG
│   ├── IMG_2510.JPG.xmp
│   ├── IMG_2510_edited (1).JPG    <----
│   └── IMG_2510_edited.JPG        <----
└── _
    ├── IMG_0096.JPG
    ├── IMG_0096.JPG.xmp
    ├── IMG_2242 (1).JPG
    ├── IMG_2242 (1).JPG.xmp
    ├── IMG_2242 (2).JPG
    ├── IMG_2242 (2).JPG.xmp
    ├── IMG_2242.JPG
    ├── IMG_2242.JPG.xmp
    ├── IMG_2242_edited (1).JPG    <----
    ├── IMG_2242_edited.JPG        <----
    ├── IMG_2768.JPG
    ├── IMG_2768.JPG.xmp
    ├── IMG_2774.JPG
    └── IMG_2774.JPG.xmp
```

It's possible the edited photos either aren't being found by osxphotos or are actually missing from the library.  Is this library synched to iCloud?  Are you using referenced photos (photos kept outside the Photos library) or the default "copy items to the Photos library"? (In `Photos > Preferences`)

Here's a command that might help figure out if the edited files are being found by osxphotos:

`osxphotos query --quiet --print "{original_name}{comma}{uuid}{comma}{edited,-}{comma}{photo.path_edited}" --name DSC_0001`

Replace `DSC_0001` with the name of a photo which is exhibiting the problem.  You'll get a custom CSV report showing whether or not the photo is edited and if it is, the path to the photo. If you find a photo that is edited but has no path that may be an indication.

Alternatively, you could do this command to print a report of all edited photos and their path:

`osxphotos query --quiet --print "{original_name}{comma}{uuid}{comma}{edited,-}{comma}{photo.path_edited}" --edited`

**Comment by drodner** [comment]
I ran the 2 experiments and here are the results. I selected 10 images, 3 of which were edited, with the same filename, DSC_0006.JPG.

Without the --filename switch, only 1 _edited image was exported and its filename did not match the corresponding original. DSC_0006_edited.JPG should have been DSC_0006 (1)_edited.JPG. So I did get the same problem. 

With the --filename switch, still only 1 of 3 _edited images was exported but this time the filename did match with the corresponding original. This was good.

Just for kicks, I did a second image, this time selecting 4 of 10 edited images. Same results. Again, only 1 edited image exported.

For the _edited images that failed to export, the error in the output text said that the Edited image was missing, exporting original. I'm pretty sure all of the edited images do exist. In Photos, when I click on the Image menu while viewing the editing image, the Revert to Original entry is enabled. I did not actually revert, but the option is enabled.

Small sample size I know, but it seems unusual that only one edited image in each test would actually export and the others would fail (all with the same Edited image missing, exporting original error.

**Reply by RhetTbull** [reply]
> For the _edited images that failed to export, the error in the output text said that the Edited image was missing

Aha! That's a clue. So the problem is that either the edited images are missing (meaning Photos can't find them) ort they are not but osxphotos cannot find them.  See the `osxphotos query` commands I posted above. Try those to collect some more data.  It might be interesting to run this on your entire library to see how many edited photos appear to be missing:

`osxphotos query --quiet --print "{original_name}{comma}{uuid}{comma}{edited,-}{comma}{photo.path_edited}" --edited`

Also, are you using the default "Copy items to Photos library" or a referenced library (where your photos are kept outside the library and managed by you)?

**Reply by drodner** [reply]
The library is not synched to iCloud. And I am using the default "Copy items to Photos library" setting.

Running both queries you mentioned gives results consistent with what we've seen from the export command:
   Query on DSC_0001 shows 4 edited images, only 1 has a path.
   Query on DSC_0006 shows 3 edited images. Again, only 1 has a path.

And the query on the whole library prints a long list. I counted the ones with paths and got 250, which is the number of _edited images I got in my initial full library export. (Total number of edited images is around 5900)

**Reply by RhetTbull** [reply]
This confirms it's not a bug in the export code though it could be a bug in osxphotos. Try selecting one of the photos that's missing but has been edited in Photos and using the Photos export feature (`File > Export`). Does that work as expected?

**Reply by RhetTbull** [reply]
Here's another debug step to try in addition to exporting a photo via the Photos menu:

In your verbose output, for those that were missing and skipped, there should be a unique ID (UUID) string in the error message which, on Mojave, looks something like: `od0fmC7NQx+ayVr+%i06XA`

`Skipping missing edited photo Pumpkins4.jpg (od0fmC7NQx+ayVr+%i06XA)`

Copy that UUID and place it in the following command:

`osxphotos debug-dump --dump photos --uuid "od0fmC7NQx+ayVr+%i06XA" > debug.txt`

Then send me the debug.txt (rturnbull+git@gmail.com) or post it here. Note: It may contain location information so if you consider that to be sensitive you can search for the lines `latitude`, `longitude` and `reverse_geolocation` and remove those.

**Reply by oPromessa** [reply]
Not trying to hijack the discussion. 

My library is all made of references files 

* I have 3603 edited pics (as per osxphotos --edited query)
* and 3327 exported files with _edited on the filename. 
* comparing the filename files coming out of the osxphotos query command for a correctly edited pic exported and an original export what osxphotos report is that it can't find the path location of the edited file and shows '_'

```bash

$ osxphotos query --quiet --print "{original_name}{comma}{uuid}{comma}{edited,-}{comma}{photo.path_edited}" --edited > edited.csv

$ grep 9D3E0182-87C9-4A58-ADCC-765C8512178B edited.csv
P2071699,9D3E0182-87C9-4A58-ADCC-765C8512178B,edited,_

$ grep 3E5A6475-0D23-4580-9CF0-90CEAEF99FE edited.csv 
P6100490,3E5A6475-0D23-4580-9CF0-90CEAEF99FEB,edited,/Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/resources/renders/3/3E5A6475-0D23-4580-9CF0-90CEAEF99FEB_1_201_a.jpeg
```

**Reply by RhetTbull** [reply]
Thanks @oPromessa As expected your edited images are in the Photos library (which is where Photos stores the edit even for referenced photos). But the fact that you too have missing edited images may indicate a bug in osxphotos that it can't find certain edited files. I wonder if it has something to do with the source of the edit? That is, if photos were edited in external editor, are they stored differently for referenced files?  

Could you  try this:

Find photo `P2071699` in your library. Verify that Photos shows the file as edited and try to export it using Photos (do not use "export unmodified original", use "export 1 photo").  What I'm trying to figure out is if Photos itself knows where the edited image is or if its just showing you the preview image.  

Another test would be to try to edit the edited photo in Photos. Does Photo show the edit screen or do you get an error about missing photo?

If Photos can find it and osxphotos can't, I'll need some additional data to figure out the problem.

**Reply by drodner** [reply]
I tried to further edit the edited photos for DSC_0001 that osxphotos did not have a path for. All 3 of the photos were able to be edited more with no problem. Also, File | Export successfully exports the edited photo for all 3 of them.

I ran the debug dump command that you mentioned on one of the photos without a path. I will send that to you in an email.

**Reply by oPromessa** [reply]
Sorry for the delay in providing answers to your questions. Here they go...
> 
> Could you try this:
> 
> Find photo `P2071699` in your library. Verify that Photos shows the file as edited and try to export it using Photos (do not use "export unmodified original", use "export 1 photo"). What I'm trying to figure out is if Photos itself knows where the edited image is or if its just showing you the preview image.
* The "export 1 photo" worked correctly.
> 
> Another test would be to try to edit the edited photo in Photos. Does Photo show the edit screen or do you get an error about missing photo?
* Hitting "Edit" it brings up the Photos Editing screen.

> If Photos can find it and osxphotos can't, I'll need some additional data to figure out the problem.

* I'll upgrade from 0.54.4 to 0.55.1 and give it a go.
* for now the `osxphotos inspect` of `0.55.1` continues not to show the Edite pic Path:
```
Path original: /Volumes/Macintosh HD/Users/Shared/Pictures/Pictures/iPhoto/2009-03-10-2126-29/P2071699.JPG
Path preview: /Users/Shared/Pictures/iPhoto Shared Library.photoslibrary/resources/derivatives/9/9D3E0182-87C9-4A58-ADCC-765C8512178B_1_202_a.jpeg
```

**Reply by RhetTbull** [reply]
@oPromessa thanks for the test data. What version of macOS are you running? I am very close to solving @drodner's original issue (see #859) but this is specific to Mojave & below. Photos uses a heuristic for the path of the photo that produces semi-random paths (presumably to avoid too many photos in any given directory) but the heuristic on Mojave and below is very different than the heuristic on Catalina and above.

**Comment by oPromessa** [comment]
I'm with Monterey version 12.6.1 in an Intel CPU. About to update to 12.6.2 ;)

Let me know if I can provide some more debug.

**Reply by RhetTbull** [reply]
Thanks. I'll send you a debug script to collect some data later this week. The fix for @drodner 's issue (which we're close to solving) only touched code that reads the Mojave database so this is a separate issue. I'll open a new issue for this.

**Comment by RhetTbull** [comment]
This issue is fixed for Mojave in [v0.55.2](https://github.com/RhetTbull/osxphotos/releases/tag/v0.55.2) thanks to a lot of data and testing by @drodner
